### DISCUSSION TITLE
iis2dh driver - I2C instance on bus okay "discovery" issue

### DISCUSSION THREAD

**Discussion by tedhavelka** [zephyrproject-rtos/zephyr#51733] [Q&A]
Hello Zephyr Forum,

I am attempting to bring up a sensor interfacing module in a Zephyr based application, one which talks to an STMicro IIS2DH accelerometer.  My target processor is NXP's LPC55S69, and the peripheral therein I am using for I2C is this microcontroller's flexcomm4 on chip serial peripheral.  This peripheral is called out in device tree source files for the board lpcxpresso55s69, and associated microcontroller.

At build time I find that while Zephyr's device tree parsing proceeds without errors or apparent warnings, there's a conditional compilation of part of a data structure in [iis2dh.h line 49](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/iis2dh/iis2dh.h#L49).  When the Zephyr toolchain, I'm assuming arm-none-eabi-gcc pre-processor reaches this in-tree source file, neither SPI nor I2C instances are ever found to have "okay" status.  And yet I've called out not one but two sensors on the I2C bus of the target microcontroller.

What's more interesting, or strange to me, is that when I copy that DT_ macro from iis2dh.h line 49 into my application sources, there is tests true and returns '1'.  I can toggle my app-side instance of `#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)` to return true or false, by commenting out and enabling corresponding DTS overlay lines for my I2C based sensors.  But this pre-processor test always returns false during compiling of iis2dh.h.

The reason this pre-processor test is important is that when it fails, the data structure for an instance of an iis2dh device as seen by Zephyr and driver routines lacks an 'i2c' data member.  Without this there's no reference to the bus on which the sensor resides.  For this the sensor is never found to be ready by calls to `device_is_ready()`.

In the past year I have used STMicro's IIS2DH Zephyr driver code to successfully configure and talk to iis2dh accelerometers, in both Zephyr 2.6.0 and Zephyr 3.1.0.  It is only with the latest migration to a newer Zephyr release that I've run into this issue, and have spent much of the past three days tracking down a root cause, and searching documentation and forums along the way for answers and sanity checks.

I've created something of a minimal Zephyr app to showcase this issue.  It is based on Zephyr 3.2.0 samples/basic/blinky project, with some additions mostly to prj.conf Kconfig settings, and to the lone source file src/main.c.  This project I've recently pushed to github.com at https://github.com/tedhavelka/iis2dh-minimal.

Is it normal that DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c) would evaluate to 'false' as compilation processes Zephyr driver sources, and during the same build it may evaluate to 'true' in application source code?  Or are there mistakes I've made in the sample app to showcase this "instance on bus" evaluation behavior?

I'm guessing there is one or more dumb errors I am making, but any insight or help the development community can offer I apprecate!

- Ted

**Comment by DerekSnell** [comment]
Hi @tedhavelka ,
I just tested building with your app and Zephyr v3.2.0, and when iis2dh.h is built, DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c) evaluates as true for me.  I am building in Windows, and my build environment is using older tools.  I am no devicetree expert, but maybe this is dependent on the build tools.  @mbolivar-nordic , I wonder if you have thoughts on this issue?  I can try updating my build environment and Zephyr SDK to the latest and see if I find a difference.  @tedhavelka , do you want to share details about your build environment so we can compare?

I cloned your app https://github.com/tedhavelka/iis2dh-minimal into my v3.2.0 tree.  I also modified iis2dh.h with these build warnings to see how this macro was evaluated at build time.
```
#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
	struct i2c_dt_spec i2c;
	#warning iis2dh.h DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c) was true
#else
	#warning iis2dh.h DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c) was false
#endif
```

Here is my build output:
```
>west build -b lpcxpresso55s69_cpu0 samples\iis2dh-minimal -d ..\..\iis2dh_LPC5569  --pristine
-- west build: making build dir c:\Temp\zRelease\iis2dh_LPC5569 pristine
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: C:/Temp/zRelease/zephyrproject/zephyr/samples/iis2dh-minimal
-- Found Python3: C:/Python39/python.exe (found suitable exact version "3.9.6") found components: Interpreter
-- Cache files will be written to: C:/Temp/zRelease/zephyrproject/zephyr/.cache
-- Zephyr version: 3.2.0 (C:/Temp/zRelease/zephyrproject/zephyr)
-- Found west (found suitable version "0.13.0", minimum required is "0.7.1")
-- Board: lpcxpresso55s69_cpu0
-- Found toolchain: gnuarmemb (C:\nxp\MCUXpressoIDE_11.5.1_7266\ide\tools)
-- Could NOT find Dtc (missing: DTC) (Required is at least version "1.4.6")
-- Found BOARD.dts: C:/Temp/zRelease/zephyrproject/zephyr/boards/arm/lpcxpresso55s69/lpcxpresso55s69_cpu0.dts
-- Found devicetree overlay: C:/Temp/zRelease/zephyrproject/zephyr/samples/iis2dh-minimal/boards/lpcxpresso55s69_cpu0.overlay
'label' is marked as deprecated in 'properties:' in C:/Temp/zRelease/zephyrproject/zephyr/dts/bindings\sensor\st,iis2dh-i2c.yaml for node /soc/peripheral@50000000/flexcomm@8a000/iis2dh@19.
-- Generated zephyr.dts: C:/Temp/zRelease/iis2dh_LPC5569/zephyr/zephyr.dts
-- Generated devicetree_generated.h: C:/Temp/zRelease/iis2dh_LPC5569/zephyr/include/generated/devicetree_generated.h
-- Including generated dts.cmake file: C:/Temp/zRelease/iis2dh_LPC5569/zephyr/dts.cmake
Parsing C:/Temp/zRelease/zephyrproject/zephyr/Kconfig
Loaded configuration 'C:/Temp/zRelease/zephyrproject/zephyr/boards/arm/lpcxpresso55s69/lpcxpresso55s69_cpu0_defconfig'
Merged configuration 'C:/Temp/zRelease/zephyrproject/zephyr/samples/iis2dh-minimal/prj.conf'
Configuration saved to 'C:/Temp/zRelease/iis2dh_LPC5569/zephyr/.config'
Kconfig header saved to 'C:/Temp/zRelease/iis2dh_LPC5569/zephyr/include/generated/autoconf.h'
-- The C compiler identification is GNU 10.3.1
-- The CXX compiler identification is GNU 10.3.1
-- The ASM compiler identification is GNU
-- Found assembler: C:/nxp/MCUXpressoIDE_11.5.1_7266/ide/tools/bin/arm-none-eabi-gcc.exe
Load components for LPC55S69:
driver_common component is included.
driver_reset component is included.
device_CMSIS component is included.
CMSIS_Include_core_cm component is included.
driver_lpc_gpio component is included.
driver_pint component is included.
driver_inputmux component is included.
driver_inputmux_connections component is included.
driver_flexcomm_i2c component is included.
driver_flexcomm component is included.
driver_flexcomm_usart component is included.
driver_lpc_iocon component is included.
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Temp/zRelease/iis2dh_LPC5569
-- west build: building application
[1/180] Generating include/generated/version.h
-- Zephyr version: 3.2.0 (C:/Temp/zRelease/zephyrproject/zephyr), build: zephyr-v3.2.0-2-g5252f2fe4394
[9/180] Building C object CMakeFiles/app.dir/src/main.c.obj
C:/Temp/zRelease/zephyrproject/zephyr/samples/iis2dh-minimal/src/main.c:21:2: warning: #warning "- DEV 1028 - app side macro test for I2C device instances returns 1, \"one or more found\"" [-Wcpp]
   21 | #warning "- DEV 1028 - app side macro test for I2C device instances returns 1, \"one or more found\""
      |  ^~~~~~~
In file included from C:/Temp/zRelease/zephyrproject/zephyr/samples/iis2dh-minimal/src/main.c:26:
c:\temp\zrelease\zephyrproject\zephyr\drivers\sensor\iis2dh\iis2dh.h:50:3: warning: #warning iis2dh.h DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c) was true [-Wcpp]
   50 |  #warning iis2dh.h DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c) was true
      |   ^~~~~~~
C:/Temp/zRelease/zephyrproject/zephyr/samples/iis2dh-minimal/src/main.c: In function 'read_iis2dh_whoami_register':
C:/Temp/zRelease/zephyrproject/zephyr/samples/iis2dh-minimal/src/main.c:86:2: warning: #warning "- DEV 1027 - DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c) returns 'true', one or more devices found 'okay' on I2C bus" [-Wcpp]
   86 | #warning "- DEV 1027 - DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c) returns 'true', one or more devices found 'okay' on I2C bus"
      |  ^~~~~~~
[170/180] Linking C executable zephyr\zephyr_pre0.elf

[174/180] Linking C executable zephyr\zephyr_pre1.elf

[180/180] Linking C executable zephyr\zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       29392 B       160 KB     17.94%
            SRAM:        4304 B       192 KB      2.19%
        USB_SRAM:          0 GB        16 KB      0.00%
        IDT_LIST:          0 GB         2 KB      0.00%
```

**Comment by DerekSnell** [comment]
Hi @tedhavelka ,
I just updated to latest Zephyr SDK, updated Python, and all the Python modules and other tools.  DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c) still evaluates as true for me in iis2dh.h when I build.  Here is my latest build output:

```
>west build -b lpcxpresso55s69_cpu0 samples\iis2dh-minimal -d ..\..\iis2dh_LPC5569  --pristine
-- west build: making build dir c:\Temp\zRelease\iis2dh_LPC5569 pristine
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: C:/Temp/zRelease/zephyrproject/zephyr/samples/iis2dh-minimal
-- Found Python3: C:/Python311/python.exe (found suitable exact version "3.11.0") found components: Interpreter
-- Cache files will be written to: C:/Temp/zRelease/zephyrproject/zephyr/.cache
-- Zephyr version: 3.2.0 (C:/Temp/zRelease/zephyrproject/zephyr)
-- Found west (found suitable version "0.14.0", minimum required is "0.7.1")
-- Board: lpcxpresso55s69_cpu0
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.15.1 (C:/Users/nxa14743/zephyr-sdk-0.15.1)
-- Found toolchain: zephyr 0.15.1 (C:/Users/nxa14743/zephyr-sdk-0.15.1)
-- Found Dtc: C:/ProgramData/chocolatey/bin/dtc.exe (found suitable version "1.5.0", minimum required is "1.4.6")
-- Found BOARD.dts: C:/Temp/zRelease/zephyrproject/zephyr/boards/arm/lpcxpresso55s69/lpcxpresso55s69_cpu0.dts
-- Found devicetree overlay: C:/Temp/zRelease/zephyrproject/zephyr/samples/iis2dh-minimal/boards/lpcxpresso55s69_cpu0.overlay
'label' is marked as deprecated in 'properties:' in C:/Temp/zRelease/zephyrproject/zephyr/dts/bindings\sensor\st,iis2dh-i2c.yaml for node /soc/peripheral@50000000/flexcomm@8a000/iis2dh@19.
-- Generated zephyr.dts: C:/Temp/zRelease/iis2dh_LPC5569/zephyr/zephyr.dts
-- Generated devicetree_generated.h: C:/Temp/zRelease/iis2dh_LPC5569/zephyr/include/generated/devicetree_generated.h
-- Including generated dts.cmake file: C:/Temp/zRelease/iis2dh_LPC5569/zephyr/dts.cmake
Parsing C:/Temp/zRelease/zephyrproject/zephyr/Kconfig
Loaded configuration 'C:/Temp/zRelease/zephyrproject/zephyr/boards/arm/lpcxpresso55s69/lpcxpresso55s69_cpu0_defconfig'
Merged configuration 'C:/Temp/zRelease/zephyrproject/zephyr/samples/iis2dh-minimal/prj.conf'
Configuration saved to 'C:/Temp/zRelease/iis2dh_LPC5569/zephyr/.config'
Kconfig header saved to 'C:/Temp/zRelease/iis2dh_LPC5569/zephyr/include/generated/autoconf.h'
-- The C compiler identification is GNU 12.1.0
-- The CXX compiler identification is GNU 12.1.0
-- The ASM compiler identification is GNU
-- Found assembler: C:/Users/nxa14743/zephyr-sdk-0.15.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc.exe
Load components for LPC55S69:
driver_common component is included.
driver_reset component is included.
device_CMSIS component is included.
CMSIS_Include_core_cm component is included.
driver_lpc_gpio component is included.
driver_pint component is included.
driver_inputmux component is included.
driver_inputmux_connections component is included.
driver_flexcomm_i2c component is included.
driver_flexcomm component is included.
driver_flexcomm_usart component is included.
driver_lpc_iocon component is included.
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Temp/zRelease/iis2dh_LPC5569
←[92m-- west build: building application
[1/180] Generating include/generated/version.h
-- Zephyr version: 3.2.0 (C:/Temp/zRelease/zephyrproject/zephyr), build: zephyr-v3.2.0-2-g5252f2fe4394
[9/180] Building C object CMakeFiles/app.dir/src/main.c.obj
C:/Temp/zRelease/zephyrproject/zephyr/samples/iis2dh-minimal/src/main.c:21:2: warning: #warning "- DEV 1028 - app side macro test for I2C device instances returns 1, \"one or more found\"" [-Wcpp]
   21 | #warning "- DEV 1028 - app side macro test for I2C device instances returns 1, \"one or more found\""
      |  ^~~~~~~
In file included from C:/Temp/zRelease/zephyrproject/zephyr/samples/iis2dh-minimal/src/main.c:26:
c:\temp\zrelease\zephyrproject\zephyr\drivers\sensor\iis2dh\iis2dh.h:50:10: warning: #warning iis2dh.h DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c) was true [-Wcpp]
   50 |         #warning iis2dh.h DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c) was true
      |          ^~~~~~~
C:/Temp/zRelease/zephyrproject/zephyr/samples/iis2dh-minimal/src/main.c: In function 'read_iis2dh_whoami_register':
C:/Temp/zRelease/zephyrproject/zephyr/samples/iis2dh-minimal/src/main.c:86:2: warning: #warning "- DEV 1027 - DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c) returns 'true', one or more devices found 'okay' on I2C bus" [-Wcpp]
   86 | #warning "- DEV 1027 - DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c) returns 'true', one or more devices found 'okay' on I2C bus"
      |  ^~~~~~~
[170/180] Linking C executable zephyr\zephyr_pre0.elf

[174/180] Linking C executable zephyr\zephyr_pre1.elf

[180/180] Linking C executable zephyr\zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       29408 B       160 KB     17.95%
            SRAM:        4304 B       192 KB      2.19%
        USB_SRAM:          0 GB        16 KB      0.00%
        IDT_LIST:          0 GB         2 KB      0.00%

```

I hope you can find the issue on your end.  Best regards

**Comment by thavelka-cornell** [comment]
Hello @DerekSnell,

Thank you for looking into my question on Zephyr's DT_ANY_INST_ON_BUS_STATUS_OKAY device tree macro.  I believe I found the issue just several hours ago, in a misplaced `DT_DRV_COMPAT st_iis2dh` stanza.  I happen to be building on an Ubuntu based host, where I've installed Zephyr build tools within the past year.  But I think we can safely say my bug was my own, and not a toolchain difference, which could be much more subtle.

During some code clean up I must have moved  `DT_DRV_COMPAT st_iis2dh` to a section where I had other pound defines, and beyond the pound include of `iis2dh.h`.  That broke the working of macro DT_ANY_INST_ON_BUS_STATUS_OKAY in that header file, and I just didn't realize what was going on because the build completed with no errors or warnings.

I was reminded of this requirement when searching for possible answers, and I found a reply from Marti Bolivar to another developer:

https://github.com/zephyrproject-rtos/zephyr/issues/41677#issuecomment-1011605859

Marti was speaking about the macro DT_INST_FOREACH_STATUS_OKAY.  This macro however seems similar in some ways to the one I found in iis2dh.h, and I am guessing when I look up the definition of DT_ANY_INST_ON_BUS_STATUS_OKAY that it somewhere depends on DT_DRV_COMPAT expanding to some non-zero value.

One thing that I am wondering, my issue first arose when both macro based tests failed silently in the building of `iis2dh.h`.  When device instances with "okay" status are not found on either bus, the IIS2DH driver compiles, but produces code which will never find a sensor that's ready to use.  Would it be helpful if zephyr/drivers/sensor/iis2dh/iis2dh.h produced a build time warning when neither SPI nor I2C buses have any okay devices on them?

- Ted

**Comment by DerekSnell** [comment]
Hi @tedhavelka ,
I'm glad you found your issue.  

> Would it be helpful if zephyr/drivers/sensor/iis2dh/iis2dh.h produced a build time warning when neither SPI nor I2C buses have any okay devices on them?

That sounds helpful to me.

Best regards
